app/services/places.js

```diff
@@ -0,0 +1,95 @@
+/*
+ Copyright 2016 Mozilla
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+import {
+  datascript,
+  mori,
+  helpers,
+} from 'datascript-mori';
+import { DB } from './user-agent-service/sqlite';
+
+const { vector, parse, conj } = mori;
+const { DB_ADD } = helpers;
+
+export class PlacesImporter {
+  constructor(placesDB) {
+    this.db = placesDB;
+    this.places = new Map();
+  }
+
+  async buildPlaceMap(limit = undefined) {
+    const placeQuery = `SELECT id, url, title, visit_count, last_visit_date, guid
+    FROM moz_places
+    WHERE hidden = 0
+    ${limit ? ' LIMIT ${limit}' : ''}
+    `;
+    await this.db.each(placeQuery, [], (row) => {
+      this.places.set(row.id, {
+        id: row.id,
+        url: row.url,
+        title: row.title,
+        visit_count: row.visit_count,
+        last_visit_date: row.last_visit_date,
+        guid: row.guid,
+        visits: [],
+      });
+    });
+
+    const visitQuery = `SELECT id, from_visit, place_id, visit_date, visit_type, session
+    FROM moz_historyvisits`;
+    await this.db.each(visitQuery, [], (row) => {
+      const place = this.places.get(row.place_id);
+      if (!place) {
+        return;
+      }
+
+      // For now, build entire places tree in memory.  TODO: stream to the DS store.
+      place.visits.push(row.visit_date);
+    });
+
+    this.places.forEach((place) => {
+      place.visits.sort((a, b) => a - b);
+    });
+
+    // TODO: sort places; prune places with no visits.
+  }
+
+  writePlacesTo(conn) {
+    const placeEntities = [];
+
+    this.places.forEach((place) => {
+      let txn = vector(vector(DB_ADD, -1, 'page/url', place.url));
```

Comment:
Also `page/title`, `place/guid` while we're here — might as well get us most of the way there.